File: frontend/src/standees/budget.test.ts
// The world-wide Standees budget (#371, ADR-0015): a user may have at most 3
// Standees out at once, counted across *every* map. This mirrors the backend
// `Standees::Budget` arithmetic so the deploy affordance can show "N of 3 out"
// and refuse — with a located pointer — before the employee writes anything.

import { describe, expect, it } from 'vitest'
import { STANDEE_CAP, standeeBudget } from './budget.ts'

const at = (map_title: string, x: number, y: number) => ({ map_title, x, y })

describe('standeeBudget', () => {
  it('with none out allows a deploy and reports the full budget', () => {
    const b = standeeBudget([])
    expect(b.out).toBe(0)
    expect(b.remaining).toBe(STANDEE_CAP)
    expect(b.allowed).toBe(true)
    expect(b.reason).toBeNull()
  })

  it('below the cap still allows, counting the remaining down', () => {
    const b = standeeBudget([at('Plaza', 3, 5), at('Grove', 2, 2)])
    expect(b.out).toBe(2)
    expect(b.remaining).toBe(1)
    expect(b.allowed).toBe(true)
    expect(b.reason).toBeNull()
  })

  it('at the cap refuses with a pointer naming where each Standee is', () => {
    const b = standeeBudget([at('Plaza', 3, 5), at('Grove', 2, 2), at('Plaza', 1, 1)])
    expect(b.out).toBe(3)
    expect(b.remaining).toBe(0)
    expect(b.allowed).toBe(false)
    expect(b.reason).toContain('Plaza (3, 5)')
    expect(b.reason).toContain('Grove (2, 2)')
    expect(b.reason).toContain('Plaza (1, 1)')
  })

  it('counts across maps — three on three maps is still the cap', () => {
    const b = standeeBudget([at('A', 0, 0), at('B', 0, 0), at('C', 0, 0)])
    expect(b.allowed).toBe(false)
    expect(b.remaining).toBe(0)
  })
})

File: backend/test/models/standees/budget_test.rb
require "test_helper"

# The Standees budget (#371, ADR-0015): a user may have at most 3 Standees out
# at once, counted across *every* map — not 3 per map. These specs pin the
# arithmetic (out / remaining / allows?) at, below and over the cap, and the
# load-bearing rule that the count spans maps rather than being per-map.
module Standees
  class BudgetTest < ActiveSupport::TestCase
    setup do
      @company, @user = setup_company
    end

    def make_map(slug:)
      ::Maps::Map.create!(
        slug: slug, title: slug.titleize, multiplayer: true,
        access_policy: { "kind" => "public" }, cols: 8, rows: 8, baked: {}
      )
    end

    def stand(map, user: @user, x: 1, y: 1, message: "hi")
      ::Standees::Standee.create!(map: map, user: user, cell_x: x, cell_y: y, message: message)
    end

    test "an employee with none out may deploy — 3 remaining" do
      budget = ::Standees::Budget.for(@user)
      assert_equal 0, budget.out
      assert_equal 3, budget.remaining
      assert budget.allows?
      assert_nil budget.refusal
    end

    test "below the cap still allows, remaining counts down" do
      map = make_map(slug: "plaza")
      stand(map)
      stand(map)

      budget = ::Standees::Budget.for(@user)
      assert_equal 2, budget.out
      assert_equal 1, budget.remaining
      assert budget.allows?
    end

    test "at the cap the deploy is refused with a located pointer" do
      plaza = make_map(slug: "plaza")
      grove = make_map(slug: "grove")
      stand(plaza, x: 3, y: 5)
      stand(grove, x: 2, y: 2)
      stand(plaza, x: 1, y: 1)

      budget = ::Standees::Budget.for(@user)
      assert_equal 3, budget.out
      assert_equal 0, budget.remaining
      refute budget.allows?
      # The refusal names where the existing Standees are standing.
      assert_match "Plaza (3, 5)", budget.refusal
      assert_match "Grove (2, 2)", budget.refusal
      assert_match "Plaza (1, 1)", budget.refusal
    end

    test "the count spans maps, not per-map" do
      # Three maps, one Standee each — that is the cap, even though no single map
      # holds more than one. Per-map counting would wrongly still allow.
      %w[a b c].each { |slug| stand(make_map(slug: slug)) }

      budget = ::Standees::Budget.for(@user)
      assert_equal 3, budget.out
      refute budget.allows?
    end

    test "another employee's Standees do not count against mine" do
      _, other = setup_company
      map = make_map(slug: "plaza")
      stand(map, user: other)
      stand(map, user: other)
      stand(map, user: other)

      budget = ::Standees::Budget.for(@user)
      assert_equal 0, budget.out
      assert budget.allows?
    end
  end
end
